source/common/listener_manager/listener_impl.cc

```diff
@@ -801,6 +921,52 @@ absl::Status ListenerImpl::buildFilterChains(const envoy::config::listener::v3::
       *filter_chain_manager_);
 }
 
+absl::Status ListenerImpl::buildFilterChains(
+    const FilterChainRefVector& added_filter_chains,
+    const absl::flat_hash_set<absl::string_view>& removed_filter_chains) {
+  transport_factory_context_->setInitManager(*dynamic_init_manager_);
+  ListenerFilterChainFactoryBuilder builder(*this, *transport_factory_context_);
+  return filter_chain_manager_->addFilterChains(
+      configInternal().has_filter_chain_matcher() ? &configInternal().filter_chain_matcher()
+                                                  : nullptr,
+      added_filter_chains, removed_filter_chains,
+      configInternal().has_default_filter_chain() ? &configInternal().default_filter_chain()
+                                                  : nullptr,
+      builder, *filter_chain_manager_);
+}
+
+void ListenerImpl::trackOriginInitDependencies(ListenerImpl& origin) {
+  if (origin.dynamic_init_manager_->state() != Init::Manager::State::Initializing) {
+    return;
+  }
+
+  // If the origin listeners are in the process of initializing, we need keep reference to the
+  // init objects so new listener will only be initialized after all previous versions dependencies
+  // are also initialized.
+  origins_dynamic_init_manager_ = origin.origins_dynamic_init_manager_;
```

Comment:
That doesn't sound right. I think a replaced initializating listener should cleanly delete itself from the initialization sequence. Is this done this way because we can't remove an init target from a manager?